lib/Dialect/Arith/Conversions/ArithToModArith/ArithToModArith.cpp

```diff
@@ -102,17 +102,30 @@ struct ConvertConstant : public OpConversionPattern<mlir::arith::ConstantOp> {
       return failure();
     }
 
-    auto denseIntAttr = dyn_cast<DenseIntElementsAttr>(op.getValue());
-    if (denseIntAttr) {
-      auto result = b.create<mod_arith::ConstantOp>(
-          typeConverter->convertType(op.getType()), denseIntAttr);
+    Type newResultType = typeConverter->convertType(op.getResult().getType());
+    mod_arith::ModArithType newResultEltTy =
+        cast<mod_arith::ModArithType>(getElementTypeOrSelf(newResultType));
+    IntegerType storageType =
+        cast<IntegerType>(newResultEltTy.getModulus().getType());
+    unsigned newWidth = storageType.getWidth();
+
+    if (auto denseIntAttr = dyn_cast<DenseIntElementsAttr>(op.getValue())) {
+      auto shapedType = cast<ShapedType>(newResultType);
+      SmallVector<APInt, 4> newValues;
+      for (const APInt &val : denseIntAttr.getValues<APInt>()) {
+        newValues.push_back(val.sextOrTrunc(newWidth));
+      }
+      TypedAttr newAttr = DenseIntElementsAttr::get(
+          RankedTensorType::get(shapedType.getShape(), storageType), newValues);
```

Comment:
Sometimes, working with MLIR is just *beautiful* xD 

Surprisingly, it seems like this is the only place where we need to do this kind of `DenseIntElementsAttr`-to-`DenseIntElementsAttr` conversion, so it's probably not worth making this a helper...

Reply:
This issue will disappear if / when I upstream some helpers related to this that will allow us to construct dense int elements attributes using mod arith types directly.